_posts/2021-07-07-backpressure.md

```diff
@@ -0,0 +1,189 @@
+  <img src="{{ site.baseurl }}/img/blog/2021-07-07-backpressure/bottleneck-zoom.png" alt="Bottleneck alternating between two tasks"/>
+	<p class="align-center">Bottleneck alternating between two tasks</p>
+</div>
+
+<div class="row front-graphic">
+  <img src="{{ site.baseurl }}/img/blog/2021-07-07-backpressure/sliding-window.png" alt="SlidingWindowOperator"/>
+	<p class="align-center">SlidingWindowOperator</p>
+</div>
+
+In this particular example, `SlidingWindowOperator` was the bottleneck as long as it was accumulating records.
+However, as soon as it starts to fire its windows (once every 10 seconds), the downstream task
+`SlidingWindowCheckMapper -> Sink: SlidingWindowCheckPrintSink` becomes the bottleneck and `SlidingWindowOperator`
+gets backpressured. As those busy/backpressured/idle metrics are averaging time over a couple of seconds,
+this subtlety is not immediately visible and has to be read between the lines. On top of that, the web UI
+is updating its state only once every 10 seconds, which makes spotting more frequent changes a bit more difficult.
+
+## What can I do with backpressure?
+
+In general this is a complex topic that is worthy of a dedicated blog post. It was, to a certain extent,
+addressed in [previous blog posts](https://flink.apache.org/2019/07/23/flink-network-stack-2.html#:~:text=this%20is%20unnecessary.-,What%20to%20do%20with%20Backpressure%3F,-Assuming%20that%20you).
+In short, there are two high-level ways of dealing with backpressure. Either add more resources (more machines,
+faster CPU, more RAM, better network, using SSDs…) or optimize usage of the resources you already have
+(optimize the code, tune the configuration, avoid data skew). In either case, you first need to analyze
+what is causing backpressure by:
+1. Identifying the presence of backpressure.
+2. Locating which subtask(s) or machines are causing it.
+3. Digging deeper into what part of the code is causing it and which resource is scarce.
+
+Backpressure monitoring improvements and metrics can help you with the first two points. To tackle the
+last one, profiling the code can be the way to go. To help with profiling, also starting from Flink 1.13,
+[Flame Graphs](http://www.brendangregg.com/flamegraphs.html) are [integrated into Flink's web UI](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/ops/debugging/flame_graphs/).
+Flame Graphs is a well known profiling tool and visualization technique and I encourage you to give it a try. 
+
+But keep in mind that after locating where the bottleneck is, you can analyze it the same way you would
+any other non-distributed application (by checking resource utilization, attaching a profiler, etc).
+Usually there is no silver bullet for problems like this. You can try to scale up but sometimes it might
+not be easy or practical to do.
+
+Anyway... The aforementioned improvements to backpressure monitoring allow us to easily detect the source of backpressure,
```

Comment:
The aforementioned improvements to backpressure monitoring allow us to easily detect the source of backpressure,
```